### ISSUE TITLE
Bootstrap without sprockets

### ISSUE THREAD

**Issue by matteeyah** [twbs/bootstrap-rubygem#247]
Hey there Bootstrap folk. The default asset pipeline for Rails 7 changed significantly. It no longer uses `sprockets` - the default asset pipeline library is now `propshaft`.

The bootstrap ruby gem has a hard dependency of `sprockets` which prevents it from being used with the `propshaft` asset pipeline (since `propshaft` is a replacement for `sprockets`, you can't run them side by side without hacking a huge part of it).

Here's some context around the switch to propshaft:

- https://twitter.com/dhh/status/1439878502526955521?lang=en
- https://world.hey.com/dhh/introducing-propshaft-ee60f4f6
- https://github.com/rails/propshaft

Is there a way to make this gem support both `propshaft` and `sprockets`, and if not, can we get a version of the gem for the new Rails 7 defaults?

---

Bonus notes: The way SASS assets are processed in the asset pipeline also changed. Instead of the in-process `sassc` it switched to using an external process `dartsass-rails` (https://twitter.com/dhh/status/1484581459277012992?lang=en and https://github.com/rails/dartsass-rails). Not sure if there's some incompatibility there as well.

**Comment by matteeyah** [comment]
@glebm What would be the lift for making this gem work with `propshaft`? Maybe I can draft up a PR with a little bit of guidance - where are the points where `sprockets` is a hard dependency for the rails bootstrap engine?

**Comment by glebm** [comment]
Can't you simply use twbs/bootstrap directly with propshaft?

**Comment by matteeyah** [comment]
Only if I'm not using [`importmaps`](https://github.com/rails/importmap-rails). When using importmaps there's no node package installation at all, all JavaScript is delivered through [JSPM](https://jspm.org/) as ES Modules.

This means I can't vendor the bootstrap code and expose it in an asset path for CSS compilation.

**Comment by matteeyah** [comment]
A small side-note: Importmaps and Propshaft are the new default for Rails applications.

They still don't have a huge market share in Rails applications because they're really new, and you can swap them out for an external JavaScript/SCSS bundling process, but over time they're going to grow in market share. Most older applications aren't running into this problem because they have a different setup, which is why there's not much demand for this - only new applications are affected, and you can sidestep the problem.

To me, it makes a lot of sense to support the latest approach (in addition to being backwards compatible) - which is why I'm motivated to help out with this.

**Comment by matteeyah** [comment]
I just found a relatively simple workaround - you can directly vendor the bootstrap styles as a gem with https://rails-assets.org/#/components/bootstrap

**Comment by tgaff** [comment]
> A small side-note: Importmaps and Propshaft are the new default for Rails applications.

Import maps are default in 7, but I think you might be mistaken about Propshaft.  

The rails/propshaft repo readme indicates its not yet the default.  https://github.com/rails/propshaft#will-propshaft-replace-sprockets-as-the-rails-default. It also mentions: "_This is very much beta software at the moment._"

That all being said, definitely worth keeping an eye on it!

**Comment by coorasse** [comment]
propshaft is no default in Rails 7, propshaft also would not compile your SASS, that would be completely delegated to css-bundling, so I don't think there's anything that this gem should do

**Comment by manuelmeurer** [comment]
@matteeyah rails-assets.org is not recommended anymore: https://github.com/tenex/rails-assets/#project-status
I've used it years ago and the website/domain was actually quite unstable, and everytime it was down, I couldn't deploy or CI my apps... 😤

**Comment by manuelmeurer** [comment]
I have just successfully set up a new Rails 7.1 app with Importmaps and Propshaft instead of Sprockets, using the CSS and JS from this gem.

Learnings:

* If you use importmaps instead of jsbundling-rails, go with dartsass-rails, not cssbundling-rails. [It allows you to skip using Node.js/Bun/Yarn entirely](https://github.com/rails/cssbundling-rails#how-does-this-compare-to-tailwindcss-rails-and-dartsass-rails), and it [automatically makes gem assets available to Sass](https://github.com/rails/dartsass-rails/#importing-assets-from-gems), which is what you need to import the CSS from this gem.
* Pin Popper and Bootstrap in your config/importmap.rb:
  ```
  pin "popper",    to: "popper.js"
  pin "bootstrap", to: "bootstrap.min.js"
  ```
  and then simply import them in your app/javascript/application.js:
  ```
  import "popper"
  import "bootstrap"
  ```
* Import the CSS in your app/assets/stylesheets/application.scss:
  ```
  @use "bootstrap";
  ```

That should be all! 😄

**Comment by walterhorstman** [comment]
@manuelmeurer This is great! I've been struggling for ages now with Rails 7 and the "assets pipeline". For a year I have a set-up working using jsbundling-rails (with esbuild) and cssbundling-rails, but it didn't feel right, because Rails supports importmaps for the JavaScript (I switched from CoffeeScript to native JavaScript for the whole project). Just like @matteeyah I want to adopt the new defaults as soon as possible.

Is it just me or is it very unclear what is happening with this new "assets pipeline"? I am migrating a Rails 4.2 project to Rails 7.1 (by starting from scratch in Rails 7.0 and migrating file-by-file) but the assets give me most difficulties. To me - using Rails since 2006 (!) - it's absolutely not "Rails-like": it's very hard to implement this.

**Comment by walterhorstman** [comment]
@manuelmeurer By the way, I've now included your Github repository in `Gemfile` like this:

`gem('bootstrap', branch: 'propshaft', git: 'https://github.com/manuelmeurer/bootstrap-rubygem')`

Is this the right way? Or it something like "freezing" gems still possible?

**Comment by matteeyah** [comment]
Looks like adding support for the new SASS processors wasn't as hard as I thought - https://github.com/twbs/bootstrap-rubygem/pull/271

**Comment by walterhorstman** [comment]
Simply adding `gem('bootstrap')` to `Gemflile` doesn't seem to work. Do I need a special branch or version?

**Comment by walterhorstman** [comment]
Another thing I've noticed in switching from **esbuild** to **importmaps** - in relation to this gem - is that I have to load the whole `bootstrap.js` instead of only certain components: in my application I only use the `Modal` component and I with **esbuild** I used `import { Modal } from 'bootstrap'` but now I need `import 'bootstrap'` and instead of using `new Modal(...)` I need to use `new bootstrap.Modal(...)`.

**Comment by Peredery** [comment]
> Simply adding `gem('bootstrap')` to `Gemfile` doesn't seem to work. Do I need a special branch or version?
> 
> I've now changed it to `gem 'bootstrap', git: 'https://github.com/twbs/bootstrap-rubygem'` and that works!

Until a new version is released on [rubygems](https://rubygems.org/gems/bootstrap) you should use from git

**Comment by walterhorstman** [comment]
Thanks @Peredery for your reply. I think I did the right thing then. ;-)

I know this is not the right place to ask for a question, but I hope somebody wants to answer anyway.

I was using esbuild, jsbundling-rails and cssbundling-rails to compile Bootstrap for RubyOnRails 7.1. But to keep things super simple and looking at advice from the Rails-contributors themselves, I decided to switch to importmaps. But it gives several issues as well, so I'm really curious to here your opinion.

Thanks in advance.
